substrate/client/src/block_builder.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility struct to build a block.
+
+use std::vec::Vec;
+use codec::{Joiner, Slicable};
+use state_machine::{self, CodeExecutor};
+use primitives::{Header, Block};
+use primitives::block::Transaction;
+use {backend, error, BlockId, BlockStatus, Client};
+use triehash::ordered_trie_root;
+
+/// Utility for building new (valid) blocks from a stream of transactions.
+pub struct BlockBuilder<B, E> where
+	B: backend::Backend,
+	E: CodeExecutor + Clone,
+	error::Error: From<<<B as backend::Backend>::State as state_machine::backend::Backend>::Error>,
+{
+	header: Header,
+	transactions: Vec<Transaction>,
+	executor: E,
+	state: B::State,
+	changes: state_machine::OverlayedChanges,
+}
+
+impl<B, E> BlockBuilder<B, E> where
+	B: backend::Backend,
+	E: CodeExecutor + Clone,
+	error::Error: From<<<B as backend::Backend>::State as state_machine::backend::Backend>::Error>,
+{
+	/// Create a new instance of builder from the given client.
+	pub fn new(client: &Client<B, E>) -> error::Result<Self> {
```

Comment:
should be able to choose which block to build upon

Reply:
why?

Reply:
The way the consensus system is designed, each authority imports a block, kicking off a BFT agreement focused around the statement table which eventually produces a block. It's something like a `fn build_next_block(just_imported: H256) -> Block`. It would be a simple source of a fault to have an authority accidentally build their proposal on some other block.

It's also convenient in general to be able to build on any block state is still maintained for.

Reply:
surely for any kind of useful block building (read: building a block that stands a chance of reaching > 2/3 votes), you would always have to build on the latest block?

Reply:
yeah, absolutely, since the consensus is still fault-tolerant, but any fault that an authority makes will increase latency to reach consensus. better to avoid it as well as we can.